### PR TITLE
Adding a class for reading data format of Glory Duck project.

### PR DESCRIPTION
**Pull request by dkerszberg**

### CODE REVIEW

File: include/GloryDuckTables2019Lkl.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+{
+ public:
+  // constructors
+  GloryDuckTables2019Lkl(TString inputString="");
+  
+  // destructor
+  virtual ~GloryDuckTables2019Lkl();
+ 
+  // make the TObjArray own its contents
+  void SetOwner(Bool_t set=kTRUE)            {fSampleArray->SetOwner(set);}
```
Comment (javierrico) [inline include/GloryDuckTables2019Lkl.h]: Why is this needed? it is not in the Fermi version.
In any case either setting ownership is needed or not, but should not be a decision of the user, should it?
I mean, if it is needed, just hard-code the call in the appropriate place, right?

Comment (dkerszberg) [reply include/GloryDuckTables2019Lkl.h]: It is not needed (at least I don't use it). I added it when I created the TObjectArray for the class (inspired by the JointLkl class) and at that time I copied all available function concerning the TObjectArray just in case I needed them later... I will delete that line.

File: include/GloryDuckTables2019Lkl.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+  // add samples
+  inline void AddSample(Lkl* sample)        {fSampleArray->Add(sample); SetChecked(kFALSE);}
+
+  // getters
+  inline const TObjArray* GetSampleArray()     const {return fSampleArray;}
+  inline       Int_t      GetNSamples()        const {return fSampleArray->GetEntries();}
+  inline       Lkl*       GetReferenceSample() const {return (Lkl*)fSampleArray->At(0);}
```
Comment (javierrico) [inline include/GloryDuckTables2019Lkl.h]: This line you don't need. In JointLkl the reference sample is the one for which units are the same as in the JointLkl, and all the rest are related to the reference by the units. Here there is no need since the samples are not part of a joint Lkl, but different alternatives depending on the value of the mass, and by construction they all have the same units

File: include/GloryDuckTables2019Lkl.h (hunk 1 above)
Comment (javierrico) [reply include/GloryDuckTables2019Lkl.h]: Ok, I need to revise the use of JointLkl::SetOwner, because my comment applies also there

File: include/GloryDuckTables2019Lkl.h [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  inline       Lkl*       GetSample(Int_t i)   const {return (Lkl*)fSampleArray->At(i);}
+  inline       Double_t   GetLogJ()            const {return fLogJ;}
```
Comment (javierrico) [inline include/GloryDuckTables2019Lkl.h]: This and fLogJ are not needed, since the values of <sv> are for some assumed value of J, that will need to be provided externally.
I'm actually wondering if we should not propose to the GD group that the data we change is lkl vs <sv>*J, so that no value of J needs to be assumed, which would make the thing less prone to mistakes (e.g assuming different J values for same object by different instruments)

File: include/GloryDuckTables2019Lkl.h [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  inline       UInt_t     GetNMasses()         const {return fNMasses;}
+  inline       UInt_t     GetActiveMass()      const {return fActiveMass;}
+
+  // unable SetUnitsOfG, dPhi/dE_signal must have proper units
+  virtual void SetUnitsOfG(Double_t unit);
```
Comment (javierrico) [inline include/GloryDuckTables2019Lkl.h]: On a second thought, the SetUnitsOfG should not be deactivated, because who knows if needed to convert e.g. to <sv>*J or <sv> in different units. I will also change that in the FermiTables class

File: include/GloryDuckTables2019Lkl.h [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+  // set mass index to the current active one
+  void SetActiveMass(Double_t mass);
```
Comment (javierrico) [inline include/GloryDuckTables2019Lkl.h]: This function should return an Int_t (0 in case of success, 1 otherwise -this is actually a general rule for functions that don't need to return anything and that can fail depending on input etc)

File: include/GloryDuckTables2019Lkl.h [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  inline  void  ClearSampleList()  {fSampleArray->Clear();}
+
+  // data input
+  Int_t CreateAndAddNewParabola(Double_t mass,Double_t svmin,Double_t svmax,Int_t npoints,Double_t* sv,Double_t* logL);
+
+ private:
+  UInt_t      fNMasses;       //   Number of masses
+  UInt_t      fActiveMass;    //   Position in the array fMass of the current active mass
+  TObjArray*  fSampleArray;   //-> Array of Lkl objects (one for each mass) 
+  Double_t*   fsvMin;         //-> [fNMasses] array with <sv> lower values
```
Comment (javierrico) [inline include/GloryDuckTables2019Lkl.h]: I think we need a list of sv values, which is valid for all masses, right? that reflects the format we decided, doesn't it?
so not this line nor the following but rather 
Double_t* fsvVals

**Review by javierrico — CHANGES_REQUESTED** [review]
I guess you should see now several comments at once, I'm not finished yet, but I am for today, so I send you this bunch and I will try to finish tomorrow

File: src/GloryDuckTables2019Lkl.cc [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+  Lkl* sample;
+  while((sample=(Lkl*)iter->Next()))
+    {
+      sample->InitMinuit();
+      sample->SetMinuitLink();
+    }
+  delete iter;
+
+  Double_t pStart[gNPars] = {(ginit==0? gRefsv :ginit)};
+  Double_t pDelta[gNPars] = {1};
```
Comment (javierrico) [inline src/GloryDuckTables2019Lkl.cc]: 1 is not the natural step for <sv>, probably something like gRefsv would be better.
Needs to be tested with realistic file

File: src/GloryDuckTables2019Lkl.cc [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+  SetParameters(gParName,pStart,pDelta);
+}  
+
+////////////////////////////////////////////////////////////////
+//
+// Check that everything is correct
+//
+// Return 0 in case of success, 1 otherwise
+//
+Int_t GloryDuckTables2019Lkl::MakeChecks()
```
Comment (javierrico) [inline src/GloryDuckTables2019Lkl.cc:153]: Add check that the different data members (fNMasses, fsv,fMass,...) exist and have sensible values (e.g. not NULL)

File: src/GloryDuckTables2019Lkl.cc [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+
+  cout << "GloryDuckTables2019Lkl::ReadGloryDuckInputData (" << GetName() << ") Message: Read " <<  GetSampleArray()->GetEntries() << " parabolas for a total of " << fNMasses << " DM masses." << endl;
+  return 0;
+}
+
+////////////////////////////////////////////////////////////////
+//
+// Add a new ParabolaLkl object representing the -2logL values vs
+// <sv> for one DM mass
+//
```
Comment (javierrico) [inline src/GloryDuckTables2019Lkl.cc:274]: Describe each of the arguments of this function

File: include/GloryDuckTables2019Lkl.h (hunk 6 above)
Comment (javierrico) [reply include/GloryDuckTables2019Lkl.h]: While revising the .cc file, I realize that the list of <sv> values is stored in every ParabolaLkl object. Then I think you do not need to keep it also here, you can always access the list of <sv> values accessing any of the parabolas.
Maybe you want to add a function to the GloryDuckTables2019 class that acts as a shortcut to access the list of <sv> values from one of the Parabolas objects (e.g. the first one).

This assumes that all parabolas have the same <sv> list, which is enforced at the definition of the format only, but not in the implementation of the class. Maybe at the level of implementation we don't want to assume it, to be more flexible in view of possible changes of format. 

Al alternative to force that all <sv> values for all Parabolas are the same is to store the list of <sv> only once at the GloryDuckTables2019 level and in the Parabola terms, in the x axis instead of <sv> you just store consecutive integers from 0 to nSvVals-1

File: src/GloryDuckTables2019Lkl.cc [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+// Return 0 in case of success, 1 otherwise
+//
+Int_t GloryDuckTables2019Lkl::CreateAndAddNewParabola(Double_t mass,Double_t svmin,Double_t svmax,Int_t npoints,Double_t* sv,Double_t* logL)
```
Comment (javierrico) [inline src/GloryDuckTables2019Lkl.cc]: Definitely, svmin and svmax are not needed, they can always be read from sv, which is stored in every parabola

File: src/GloryDuckTables2019Lkl.cc [hunk 11]
```diff
@@ -0,0 +1,403 @@
...
+  SetChecked(kFALSE);
+  return 0;
+}
+
+//////////////////////////////////////////////////////////////////
+//
+// Function called by Lkl::PrintOverview
+// Print info about the experimental data
+//
+void GloryDuckTables2019Lkl::PrintData(Int_t level)
```
Comment (javierrico) [inline src/GloryDuckTables2019Lkl.cc:310]: Print also all parabolas contents. Maybe later can be removed if we get too messy logs, but for test purposes it will be useful

File: src/GloryDuckTables2019Lkl.cc [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+// 
+// Return a pointer to the canvas, in case you want to modify it
+//
+TCanvas* GloryDuckTables2019Lkl::PlotInputData() 
+{
+  const Int_t nlines = 5;  // number of lines in parabolas canvas
+  Int_t ncols = TMath::Ceil(fNMasses/Float_t(nlines)); // number of columns in parabolas canvas
+
+  // create and divide canvas
+  TCanvas* canvas = new TCanvas("inputDataCanvas","-2logL value per energy bin and energy flux",ncols*250,nlines*250);
```
Comment (javierrico) [inline src/GloryDuckTables2019Lkl.cc]: Fix title of the canvas

File: src/GloryDuckTables2019Lkl.cc [hunk 13]
```diff
@@ -0,0 +1,403 @@
...
+  return canvas;
+}
+
+////////////////////////////////////////////////////////////////////////
+//
+// Set mass index to the current active one
+//
+void GloryDuckTables2019Lkl::SetActiveMass(Double_t mass)
+{
+  for (Int_t counter = 0; counter < fNMasses; counter++)
```
Comment (javierrico) [inline src/GloryDuckTables2019Lkl.cc:393]: Add comment briefly describing what you are checking

File: src/GloryDuckTables2019Lkl.cc [hunk 14]
```diff
@@ -0,0 +1,403 @@
...
+      gPad->Update();
+    }
+  return canvas;
+}
+
+////////////////////////////////////////////////////////////////////////
+//
+// Set mass index to the current active one
+//
+void GloryDuckTables2019Lkl::SetActiveMass(Double_t mass)
```
Comment (javierrico) [inline src/GloryDuckTables2019Lkl.cc]: We need another function that can have the same name, but takes some Int_t index as input, and sets the active mass to fMass[index]. Only check needed is that index is non negative and  <fNMasses

File: src/GloryDuckTables2019Lkl.cc [hunk 15]
```diff
@@ -0,0 +1,403 @@
...
+  f = 0;
+
+  // get internal object
+  GloryDuckTables2019Lkl* mylkl = dynamic_cast<GloryDuckTables2019Lkl*>(minuit->GetObjectFit());
+
+  // g is the global scale for dPhidESignal
+  Double_t g = par[0];
+
+  // get likelihood for current active mass
+  Lkl* sample = mylkl->GetSample(mylkl->GetActiveMass());
```
Comment (javierrico) [inline src/GloryDuckTables2019Lkl.cc]: This won't work, because sample will be interpreted as a Lkl object, not as ParabolaLkl one, use the dynamic cast as in the FermiTables2016Lkl case: 
ParabolaLkl* sample = dynamic_cast<ParabolaLkl*>mylkl->GetSample(mylkl->GetActiveMass());

**Review by javierrico — CHANGES_REQUESTED** [review]
I think this completes my review. Now I wait for you to address all comments before I have another look.

**Comment by dkerszberg** [comment]
I committed all the changes according to your review except for the one concerning SetUnitsOfG() because so far we didn't change the data format.

File: include/GloryDuckTables2019Lkl.h (hunk 4 above)
Comment (javierrico) [reply include/GloryDuckTables2019Lkl.h]: I see you did not do this because "so far we didn't change the data format", which I think is not a good reason. Even for a data format consisting on <sv> values the units are not univoque, so in that sense I think the active version of SetUnitsOfG is needed

File: include/GloryDuckTables2019Lkl.h (hunk 6 above)
Comment (javierrico) [reply include/GloryDuckTables2019Lkl.h]: I think that you do not need the fsvVals either, because those values are stored in every Parabola, is there any need to have them again saved as data members of GloryDuck class? Probably it's enough to add a function GetSvVals that access them from one of the Parabolas (e.g. the first one).

**Comment by javierrico** [comment]
A couple more of details and we're ready to go...

**Review by javierrico — APPROVED** [review]

**Comment by javierrico** [comment]
Pulled
